### PR TITLE
Fixes for functions

### PR DESCRIPTION
**Pull request by majosm**
Adds fixes needed to make outlining of inviscid/viscous fluxes work in mirgecom.

Note: Only works if functions are inlined at the beginning of `_dag_to_compiled_func` in grudge's lazy function caller, due to interactions with axis tagging (see #470).

### CODE REVIEW

File: pytato/analysis/__init__.py [hunk 1]
```diff
@@ -358,6 +367,11 @@ def map_distributed_send_ref_holder(self,
                                         ) -> FrozenSet[Array]:
         return frozenset([expr.passthrough_data])
 
+    def map_named_call_result(self, expr: NamedCallResult) -> FrozenSet[Array]:
+        # Not sure if this should descend into the function definition or not?
+        return frozenset(expr.call.bindings.values())
```
Comment (inducer) [inline pytato/analysis/__init__.py:372]: Refuse to guess, for now.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -53,7 +53,7 @@ jobs:
                 curl -L -O https://tiker.net/ci-support-v0
                 . ./ci-support-v0
                 build_py_project_in_conda_env
-                python -m pip install mypy
+                python -m pip install mypy==1.6.1
```
Comment (inducer) [inline .github/workflows/ci.yml:56]: Revert before merge

File: pytato/transform/__init__.py [hunk 3]
```diff
@@ -251,13 +254,23 @@ def __call__(self,  # type: ignore[override]
...
-    def clone_for_callee(self: _SelfMapper) -> _SelfMapper:
+    def clone_for_callee(
+            self: _SelfMapper, function: FunctionDefinition) -> _SelfMapper:
         """
         Called to clone *self* before starting traversal of a
         :class:`pytato.function.FunctionDefinition`.
         """
         return type(self)()
 
+    def update_from_callee_clone(
```
Comment (inducer) [inline pytato/transform/__init__.py:265]: Maybe shuffle the `update_from_callee_clone` changes into a separate PR?

File: pytato/transform/__init__.py [hunk 4]
```diff
@@ -907,13 +921,9 @@ def map_function_definition(self, expr: FunctionDefinition
                                                  for ret in expr.returns.values()])
         result: Set[InputArgumentBase] = set()
         for inp in all_callee_inputs:
-            if isinstance(inp, Placeholder):
-                if inp.name in expr.parameters:
-                    # drop, reference to argument
-                    pass
-                else:
-                    raise ValueError("function definition refers to non-argument "
```
Comment (inducer) [inline pytato/transform/__init__.py:915]: This error needs to stay I think. But we could add an additional interface to `outline` that "silently" introduces arguments for symbols used from the surrounding code. Probably not willy-nilly, but from a given list.

File: pytato/transform/__init__.py
Comment (inducer) [inline pytato/transform/__init__.py:1]: Just don't attempt to handle functions in MPMS. (Make sure to raise an error.)

File: pytato/transform/calls.py [hunk 5]
```diff
@@ -50,7 +50,10 @@ def __init__(self, substitutions: Mapping[str, Array]) -> None:
         self.substitutions = substitutions
 
     def map_placeholder(self, expr: Placeholder) -> Array:
-        return self.substitutions[expr.name]
+        if expr.name in self.substitutions:
```
Comment (inducer) [inline pytato/transform/calls.py:53]: Back out.

File: pytato/transform/calls.py [hunk 6]
```diff
@@ -63,20 +66,20 @@ def map_call(self, expr: Call) -> AbstractResultWithNamedArrays:
         assert isinstance(new_expr, Call)
 
         if expr.tags_of_type(InlineCallTag):
-            substitutor = PlaceholderSubstitutor(expr.bindings)
+            substitutor = PlaceholderSubstitutor(new_expr.bindings)
```
Comment (inducer) [inline pytato/transform/calls.py:69]: I agree with the change, but I'd like to make sure I trust my understanding.

I'm surprised you ran into an issue with it and would like to know about the failure you saw.

File: pytato/transform/calls.py [hunk 7]
```diff
@@ -63,20 +66,20 @@ def map_call(self, expr: Call) -> AbstractResultWithNamedArrays:
...
+                tags=new_expr.tags
             )
         else:
             return new_expr
 
     def map_named_call_result(self, expr: NamedCallResult) -> Array:
         new_call = self.rec(expr._container)
         assert isinstance(new_call, AbstractResultWithNamedArrays)
-        return new_call[expr.name]
+        return new_call[expr.name].expr
```
Comment (inducer) [inline pytato/transform/calls.py:82]: This code probably needs to check whether `new_call` (fix naming!) actually is a call or a `DictOfNamedArrays`.

File: pytato/codegen.py [hunk 8]
```diff
@@ -112,8 +113,16 @@ def __init__(self, target: Target,
         self.target = target
         self.kernels_seen: Dict[str, lp.LoopKernel] = kernels_seen or {}
 
-    def clone_for_callee(self) -> CodeGenPreprocessor:
-        return CodeGenPreprocessor(self.target, self.kernels_seen)
+    def clone_for_callee(
+            self: _SelfMapper, function: FunctionDefinition) -> CodeGenPreprocessor:
+        # Would need to sync bound_arguments/var_name_gen from clones
```
Comment (inducer) [inline pytato/codegen.py:118]: Disagree with the comment.

File: pytato/codegen.py [hunk 9]
```diff
@@ -112,8 +113,16 @@ def __init__(self, target: Target,
...
+        raise NotImplementedError("CodeGenPreprocessor does not support functions.")
+
+    def update_from_callee_clone(
+            self: _SelfMapper, function: FunctionDefinition,
+            callee_clone: _SelfMapper) -> None:
+        # Would need to sync bound_arguments/var_name_gen from clones
```
Comment (inducer) [inline pytato/codegen.py:124]: Disagree with the comment.

File: pytato/function.py [hunk 10]
```diff
@@ -132,19 +131,19 @@ class FunctionDefinition(Taggable):
...
-        from functools import reduce
 
         mapper = InputGatherer()
 
-        all_input_args: FrozenSet[InputArgumentBase] = reduce(
-            frozenset.union,
-            (mapper(ary) for ary in self.returns.values()),
-            frozenset()
-        )
+        all_placeholder_args: FrozenSet[Placeholder] = frozenset()
```
Comment (inducer) [inline pytato/function.py:137]: Back out.

File: pytato/distributed/partition.py [hunk 11]
```diff
@@ -288,8 +299,33 @@ def __init__(self,
...
+            # duplicating placeholders
+            self._placeholder_generator)
+
+    def update_from_callee_clone(
+            self,
+            function: FunctionDefinition,
+            callee_clone: _DistributedInputReplacer) -> None:
+        super().update_from_callee_clone(function, callee_clone)
+
+        self.user_input_names |= (
```
Comment (inducer) [inline pytato/distributed/partition.py:326]: This also seems rooted in wanting to allow names other than parameters in functions, which I think is not what we want.

File: pytato/transform/calls.py (hunk 6 above)
Comment (majosm) [reply pytato/transform/calls.py:66]: I don't recall if I actually ran into an issue with this. But if I understand correctly: if any of the bindings contain calls themselves, `new_expr.bindings` will have them inlined (where tagged), whereas `expr.bindings` won't.

**Review by inducer — APPROVED** [review]
:tada: Thanks!

**Comment by inducer** [comment]
@majosm The mypy fixes are in, as is #474. Could you update this so we can merge it?

**Comment by majosm** [comment]
@inducer Updated and removed mypy version change. Should be good to go.

**Comment by inducer** [comment]
Thx!

**Pull request merged**